enterprise/backend/src/metabase_enterprise/remote_sync/source/git.clj

```diff
@@ -0,0 +1,289 @@
+   [metabase.util.log :as log])
+  (:import
+   (org.apache.commons.io FileUtils)
+   (org.eclipse.jgit.api Git GitCommand TransportCommand)
+   (org.eclipse.jgit.dircache DirCache DirCacheEntry)
+   (org.eclipse.jgit.lib CommitBuilder Constants FileMode PersonIdent Ref)
+   (org.eclipse.jgit.revwalk RevCommit RevWalk)
+   (org.eclipse.jgit.transport PushResult RefSpec RemoteRefUpdate
+                               RemoteRefUpdate$Status UsernamePasswordCredentialsProvider)
+   (org.eclipse.jgit.treewalk TreeWalk)))
+
+(set! *warn-on-reflection* true)
+
+(defn- call-command [^GitCommand command]
+  (let [analytics-labels {:operation (-> command .getClass .getName) :remote false}]
+    (analytics/inc! :metabase-remote-sync/git-operations analytics-labels)
+
+    (try
+      (.call command)
+      (catch Exception e
+        (analytics/inc! :metabase-remote-sync/git-operations-failed analytics-labels)
+        (throw (ex-info (format "Git %s failed: %s" (:operation analytics-labels) (.getMessage e))
+                        analytics-labels e))))))
+
+(defn- call-remote-command [^TransportCommand command {:keys [^String token]}]
+  (let [analytics-labels {:operation (-> command .getClass .getName) :remote true}
+        credentials-provider (when token (UsernamePasswordCredentialsProvider. "x-access-token" token))]
+    (analytics/inc! :metabase-remote-sync/git-operations analytics-labels)
+
+    (try
+      (-> command
+          (.setCredentialsProvider credentials-provider)
+          (.call))
+      (catch Exception e
+        (analytics/inc! :metabase-remote-sync/git-operations-failed analytics-labels)
+        (throw (ex-info (format "Git %s failed: %s" (-> command .getClass .getName) (.getMessage e))
+                        analytics-labels e))))))
+
+(defn- qualify-branch [branch]
+  (if (str/starts-with? branch "refs/heads/")
+    branch
+    (str "refs/heads/" branch)))
+
+(defn fetch!
+  "Call fetch on the repository."
+  [{:keys [^Git git] :as git-source}]
+  (log/info "Fetching repository" {:repo (str git)})
+  (u/prog1 (call-remote-command (.fetch git) git-source))
+  (log/info "Successfully fetched repository" {:repo (str git)}))
+
+(defn- existing-git-repo [^java.io.File dir {:keys [^String token]}]
+  (io/make-parents dir)
+  (try
+    (when (.exists dir)
+      (u/prog1 (Git/open dir)
+        (fetch! {:git <> :token token})))
+    (catch Exception e
+      (log/warnf e "Existing git repo at %s is not configured correctly. Deleting it" dir)
+      (FileUtils/deleteDirectory dir)
+      nil)))
+
+(defn clone-repository!
+  "Clone git repository to a temporary directory using jgit. If the temp directory already exists, use it rather than re-clone.
+  Returns the path to the cloned `jgit.Repository` bare-repo object."
+  [{:keys [^String url ^String token] :as args}]
+  (let [dir (io/file (System/getProperty "java.io.tmpdir") "metabase-git" (-> (str/join ":" [url token]) buddy-hash/sha1 codecs/bytes->hex))
```

Comment:
Since the token is sensitive should we use a stronger hash here than sha1? Obviously it should be difficult to reverse it since a token is already random, but want to make sure we don't leak it. 

Slash it would maybe be better to not use the token for this?